Gameplay feature 5 upgrading weapons and healing henry

HYBleek opened the pull request:
I wrote a little bit of code for the third layer, I hope you can check it for me.
My codes might have many small bugs...... I did some simple tests, and it should run. I hope my code can help your discussion. If you find problems, please tell me, and I will fix them ASAP.
I may have taken up a bit of code written by my teammates because my part is entangled with others a lot.
I'm very sorry that I can't come to today's tut because of fever. Due to my illness, it may be a few days before I can start debugging.

stalight commented:
In general add JavaDocs, also if you are using other people's class as placeholders, don't put them in the pull request, also your JavaDocs are wrong, check the entity layers because those have JavaDocs implemented correctly

UTDQi reviewed (commented):
1.Consider wirtting the texts outside instead of hardcoding them.
e.g.:
class Heal {
    String youCanHealText = "You have Essence %d/%d. \n You have Artifact %d/%d \nYou can heal! \n Healing? [Y]/[N]";

    private String info_writer(int Essence_have, int Essence_need, int Artifact_have, int Artifact_need) {
        if (able_to_heal(Essence_have, Artifact_have, Essence_need, Artifact_need)) {
            return String.format(youCanHealText, Essence_have,Artifact_have,Essence_need,Artifact_need);
        }else{
            ...
        }
    }
}

By the way, it's\n instead of /n.

2. For healing.update() and upgrading.update(), I think you don't need to test if the Event is NoEvent, and using == for String may not be safe.

HYBleek commented:
> 1.Consider wirtting the texts outside instead of hardcoding them. e.g.: class Heal { String youCanHealText = "You have Essence %d/%d. \n You have Artifact %d/%d \nYou can heal! \n Healing? [Y]/[N]";
> 
> ```
> private String info_writer(int Essence_have, int Essence_need, int Artifact_have, int Artifact_need) {
>     if (able_to_heal(Essence_have, Artifact_have, Essence_need, Artifact_need)) {
>         return String.format(youCanHealText, Essence_have,Essence_need,Artifact_have,Artifact_need);
>     }else{
>         ...
>     }
> }
> ```
> 
> }
> 
> By the way, it's\n instead of /n.
> 
> 2. For healing.update() and upgrading.update(), I think you don't need to test if the Event is NoEvent, and using == for String may not be safe.

For using %d in the string. As the String I write is in the usecase level, it need to be printed at the presenter level through the interface. Hence, I can only give the interface a string. If you think I should use this format, then I will try to add interfaces and methods in the presenter. Then we will put the string-building issue at the presenter level and directly print it on the presenter.
For /n and \n, fixed
For update need to be in the no event. It is basically for preventing the player healing when fight and causing bugs. I remember we will let the player in no event if there is nothing happen.
For ==, fixed

UTDQi commented:
> 

String.format(youCanHealText, Essence_have,Essence_need,Artifact_have,Artifact_need) is a String.

HYBleek commented:
> > 
> 
> String.format(youCanHealText, Essence_have,Essence_need,Artifact_have,Artifact_need) is a String.

Got it. fixed. But I still think the string work can only be on the presenters. What about your idea?

HYBleek commented:
I try to compress the upgrade part. But it is still a lot, I will separate it into two once avaible

HYBleek commented:
Problems so far fixed

villanueva-j commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * The method will return the player decide to upgrade sword or armor.
+     * @param EquipmentChoice
+     * @return return the choice keyword [A][S]
```
Perhaps we should change the key for sword to just [W] for weapons so that it can be generalized to different types of weapons
